dipy/tracking/streamlinespeed.pyx

```diff
@@ -301,3 +301,174 @@ def set_number_of_points(streamlines, nb_points=3):
+    norm2 = 0.0
+    for d in range(D):
+        dn = streamline[next, d]-streamline[prev, d]
+        norm2 += dn*dn
+    norm2 = sqrt(norm2)
+
+    return norm1 / norm2
+
+
+cdef np.npy_intp c_compress_streamline(Streamline streamline, double tol_error, Streamline out) nogil:
+    cdef:
+        np.npy_intp N = streamline.shape[0]
+        np.npy_intp D = streamline.shape[1]
+        np.npy_intp nb_points = 0
+        np.npy_intp last_added_point_idx = 0
+        np.npy_intp last_success = 0
+        np.npy_intp last_added_point = 0
+        np.npy_intp i, k, prev, next, curr
+        double dn, in_between_dist
+
+    for i in range(N):
+        # Euclidean distance between last added point and current point.
+        in_between_dist = 0.0
+        for d in range(D):
+            dn = streamline[i, d] - streamline[last_added_point, d]
+            in_between_dist += dn*dn
+
+        in_between_dist = sqrt(in_between_dist)
+
+        if i == 0:  # First point is always added.
+            for d in range(D):
+                out[nb_points, d] = streamline[i, d]
+
+            last_success = i
+            nb_points += 1
+        elif i < N-1:  # Perform linearization if needed.
+            prev = last_added_point
+            next = i+1
+
+            # Check that each point is not offset by more than `tol_error` mm.
+            for k in range(last_added_point, i):
+                curr = k
+                dist = c_dist_to_line(streamline, prev, next, curr)
+
+                # If the current point's offset is greater than `tol_error`, use the latest success.
+                if dist > tol_error or in_between_dist > 10:
+                    for d in range(D):
+                        out[nb_points, d] = streamline[last_success, d]
+
+                    nb_points += 1
+                    last_added_point = i-1
+                    last_success = i-1
+                    break
+                # The current point's offset is ok, check the next point.
+                else:
+                    last_success = i
+
+        else:  # Last point is always added.
+            for d in range(D):
+                out[nb_points, d] = streamline[i, d]
+
+            nb_points += 1
+
+    return nb_points
+
+
+def compress_streamlines(streamlines, tol_error=0.5):
+    """ Compress streamlines by linearizing some part of them.
+
+    Basically, it consists in merging consecutive segments that are
+    near collinear. The merging is achieved by removing the point the two
+    segments have in common. The linearization process [Presseau15]_. will
+    not remove a point if it causes either an offset of more than
+    `tol_error`mm or a distance between two consecutive points to be more
+    than 10mm [Rheault15]_.
+
+    Parameters
+    ----------
+    streamlines : one or a list of array-like of shape (N,3)
+        Array representing x,y,z of N points in a streamline
+    tol_error : float (optional)
+        Tolerance error in mm. Default is 0.5mm.
```

Comment:
I would not use 0.5 mm as a default. This is too aggressive. We can ask Francois Rheault opinion but I would use 0.1 as a default.

I would also add a comment that recommendations can be found in Presseau et al. But, a general rule of thumb is that 0.01 for deterministic streamline as input and 0.1 for proba are good choices.